apps/emqx_management/src/emqx_mgmt_api_clients.erl

```diff
@@ -890,6 +945,218 @@ list_clients(QString) ->
+            format_results(Acc, NewCursor);
+        false ->
+            do_list_clients_v2(Nodes, NewCursor, QString0, Acc)
+    end;
+do_list_clients_v2(Nodes, _Cursor = #{type := ?CURSOR_TYPE_DS, iterator := Iter0}, QString0, Acc0) ->
+    #{limit := Limit} = Acc0,
+    {Rows0, Iter} = emqx_persistent_session_ds_state:session_iterator_next(Iter0, Limit),
+    NewCursor = next_ds_cursor(Iter),
+    Rows1 = drop_live_and_expired(Rows0),
+    Rows = maybe_run_fuzzy_filter(Rows1, QString0),
+    Acc1 = maps:update_with(rows, fun(Rs) -> [{undefined, Rows} | Rs] end, Acc0),
+    Acc = #{n := N} = maps:update_with(n, fun(N) -> N + length(Rows) end, Acc1),
+    case N >= Limit of
+        true ->
+            format_results(Acc, NewCursor);
+        false ->
+            do_list_clients_v2(Nodes, NewCursor, QString0, Acc)
+    end.
+
+format_results(Acc, Cursor) ->
+    #{
+        rows := NodeRows,
+        n := N
+    } = Acc,
+    Meta =
+        case Cursor of
+            done ->
+                #{
+                    hasnext => false,
+                    count => N
+                };
+            _ ->
+                #{
+                    hasnext => true,
+                    count => N,
+                    cursor => serialize_cursor(Cursor)
+                }
+        end,
+    Resp = #{
+        meta => Meta,
+        data => [
+            format_channel_info(Node, Row)
+         || {Node, Rows} <- NodeRows,
+            Row <- Rows
+        ]
+    },
+    ?OK(Resp).
+
+do_ets_select(Nodes, QString0, #{node := Node, node_idx := NodeIdx, cont := Cont} = _Cursor) ->
+    {_, QString1} = emqx_mgmt_api:parse_qstring(QString0, ?CLIENT_QSCHEMA),
+    Limit = maps:get(<<"limit">>, QString0, 10),
+    {Rows, #{cont := NewCont, node_idx := NewNodeIdx}} = ets_select(
+        QString1, Limit, Node, NodeIdx, Cont
+    ),
+    {Rows, next_ets_cursor(Nodes, NewNodeIdx, NewCont)}.
+
+maybe_run_fuzzy_filter(Rows, QString0) ->
+    {_, {_, FuzzyQString}} = emqx_mgmt_api:parse_qstring(QString0, ?CLIENT_QSCHEMA),
+    FuzzyFilterFn = fuzzy_filter_fun(FuzzyQString),
+    case FuzzyFilterFn of
+        undefined ->
+            Rows;
+        {Fn, Args} ->
+            lists:filter(
+                fun(E) -> erlang:apply(Fn, [E | Args]) end,
+                Rows
+            )
+    end.
+
+initial_ets_cursor([Node | _Rest] = _Nodes) ->
+    #{
+        type => ?CURSOR_TYPE_ETS,
+        node => Node,
+        node_idx => 1,
+        cont => undefined
+    }.
+
+initial_ds_cursor() ->
+    case emqx_persistent_message:is_persistence_enabled() of
+        true ->
+            #{
+                type => ?CURSOR_TYPE_DS,
+                iterator => init_persistent_session_iterator()
+            };
+        false ->
+            done
+    end.
+
+next_ets_cursor(Nodes, NodeIdx, Cont) ->
+    case NodeIdx > length(Nodes) of
+        true ->
+            initial_ds_cursor();
+        false ->
+            Node = lists:nth(NodeIdx, Nodes),
```

Comment:
Do nodes preserve count across requests?

Reply:
No, if the node list changes between requests, the total returned list of clients will change.  But this is the case with the current API as well, IIUC: if the client is asking for page 1, page 2, ..., and then the node list changes before it asks for, say, page 3, it'll also get strange results.

One initial idea I had was to encode the current node and the remaining nodes in the cursor.  But that introduces the problem of having to deal with the change in the cluster in a different manner:  what to do if one of the next nodes is off?  Or a new node is introduced but not in the `next` list?  Since the old API seems to suffer from this in a similar fashion, I left this idea out.  Which also makes the cursor slimmer when serializing it in the query params.  :sweat_smile:

Reply:
Maybe protect from the case when `NodeId > length(Nodes)`? Just end the iteration, instead of crash?